### PR TITLE
libimagentrycategory/init

### LINKED ISSUE #939: libimagentrycategory
A nice thing would be a library to add a "category" header field to an entry.

```toml
[entry]
category = "food"
```

### Desired Features

- [x] CRUD functionality

### Possible features to be implemented later

- [ ] Find all categories of all entries which have one, compare the textual difference and suggest addaptions (spelling correction etc) interactively.

---

I would go with the `entry` namespace in the header here, as we're probably developing even more such extension libraries for the `FileLockEntry` type, so we can keep these generic things in one namespace.

### PR DESCRIPTION
**Pull request by matthiasbeyer**
Because we might need this.

Closes #939

### CODE REVIEW

**Comment by matthiasbeyer** [comment]
@TheNeikos @asuivelentine @mario-kr review pls!

**Comment by matthiasbeyer** [comment]
btw, here we see the awesomeness of abstractions! With `toml-query` and `libimagerror` this library is only 200 lines of code, ... much less than if we'd need to write all the things by hand!

File: libimagentrycategory/src/category.rs [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+        self.get_header_mut()
+            .insert(&String::from("category.value"), Value::String(s.into()))
+            .map_err_into(CEK::HeaderWriteError)
+            .map(|_| ())
+    }
+
+    fn get_category(&self) -> Result<Option<Category>> {
+        let res = self.get_header().read(&String::from("category.value"));
+        if res.is_err() {
+            let res = res.unwrap_err();
```
Comment (asuivelentine) [inline libimagentrycategory/src/category.rs]: Why you don't use if let?

Comment (matthiasbeyer) [inline libimagentrycategory/src/category.rs]: Because I'm dump. Thanks for the good catch!

**Comment by matthiasbeyer** [comment]
Some nice ideas:

A category-register where categories are created by the user and the library checks before applying a category to an entry.

For this, the library should be like this:

* Creating a `CategorySetter` object, which would set a category "foo"
* The object checks whether this is a registered category, if not checks for similar categories (optionally of course) and proposes them to the user of the lib
* The user of the lib can now decide whether to do this or not and `entry.set_category(&category)` or whether to drop the Category object

* Reading from an Entry should be `entry.get_category() -> Result<Option<Category>>`

**Comment by matthiasbeyer** [comment]
The above commits can now be used as a base to implement:

> For this, the library should be like this:
> 
> Creating a CategorySetter object, which would set a category "foo"
> 
> The object checks whether this is a registered category, if not checks for similar categories (optionally of course) and proposes them to the user of the lib
> 
> The user of the lib can now decide whether to do this or not (CategoryBuilder.build() -> Category) and entry.set_category(&category) or whether to drop the Category object
> 
> Reading from an Entry should be entry.get_category() -> Result<Option<Category>>

**Comment by matthiasbeyer** [comment]
I am not sure whether the above idea gives us any improvements over the current interface.

**Pull request merged**